improvement: optimization of readmap function

thomas-dg opened the pull request:
## Description
There are two optimization points for the readmap function:
1. Map key cannot be an incomparable type, otherwise it will cause a panic, but the function does not distinguish 
2. When the key is an incomparable type, adjusted to fmt.Sprint(k) to avoid unreadable output

kenhuuu commented:
Thanks for contributing.

Your PR seems to be causing a failure with the feature test infrastructure. Check the Github Actions logs for more information but I've copied the stack trace here for you.

```
panic: reflect: call of reflect.Value.Elem on string Value [recovered]
panic: reflect: call of reflect.Value.Elem on string Value

goroutine 6 [running]:
testing.tRunner.func1.2({0xe68d60, 0xc0005b1860})
	/usr/local/go/src/testing/testing.go:1545 +0x3f7
testing.tRunner.func1()
	/usr/local/go/src/testing/testing.go:1548 +0x716
panic({0xe68d60?, 0xc0005b1860?})
	/usr/local/go/src/runtime/panic.go:920 +0x270
reflect.Value.Elem({0xe4dec0?, 0xc000917c70?, 0x2?})
	/usr/local/go/src/reflect/value.go:1271 +0x45f
github.com/apache/tinkerpop/gremlin-go/v3/driver/cucumber.getEdges(0xc000032053?)
	/go_app/driver/cucumber/cucumberWorld.go:180 +0xe99
github.com/apache/tinkerpop/gremlin-go/v3/driver/cucumber.(*CucumberWorld).loadAllDataGraph(0xc0002bf4a0)
	/go_app/driver/cucumber/cucumberWorld.go:115 +0x3c9
github.com/apache/tinkerpop/gremlin-go/v3/driver/cucumber.InitializeTestSuite.func1()
	/go_app/driver/cucumber/cucumberSteps_test.go:878 +0x45
github.com/cucumber/godog.(*runner).concurrent(0xc000300a80, 0x1)
	/go/pkg/mod/github.com/cucumber/godog@v0.14.1/run.go:82 +0x602
github.com/cucumber/godog.runWithOptions({_, _}, {0x0, 0x0, 0x0, {0x0, 0x0}, 0x0, {0x0, 0x0, ...}, ...}, ...)
	/go/pkg/mod/github.com/cucumber/godog@v0.14.1/run.go:284 +0x13ea
github.com/cucumber/godog.TestSuite.Run({{0x0, 0x0}, 0xf40478, 0xf40460, 0xc0002ca0b0})
	/go/pkg/mod/github.com/cucumber/godog@v0.14.1/run.go:348 +0x2d4
github.com/apache/tinkerpop/gremlin-go/v3/driver/cucumber.TestCucumberFeatures(0xc00015e9c0)
	/go_app/driver/cucumber/cucumberSteps_test.go:951 +0x347
testing.tRunner(0xc00015e9c0, 0xf40300)
	/usr/local/go/src/testing/testing.go:1595 +0x262
created by testing.(*T).Run in goroutine 1
	/usr/local/go/src/testing/testing.go:1648 +0x846
```

Also, you should add an entry to the CHANGELOG.asciidoc.

I'm not that familiar with Go, but I'll have another committer that is take a look at this as well.

thomas-dg commented:
> Thanks for contributing.
> 
> Your PR seems to be causing a failure with the feature test infrastructure. Check the Github Actions logs for more information but I've copied the stack trace here for you.
> 
> ```
> panic: reflect: call of reflect.Value.Elem on string Value [recovered]
> panic: reflect: call of reflect.Value.Elem on string Value
> 
> goroutine 6 [running]:
> testing.tRunner.func1.2({0xe68d60, 0xc0005b1860})
> 	/usr/local/go/src/testing/testing.go:1545 +0x3f7
> testing.tRunner.func1()
> 	/usr/local/go/src/testing/testing.go:1548 +0x716
> panic({0xe68d60?, 0xc0005b1860?})
> 	/usr/local/go/src/runtime/panic.go:920 +0x270
> reflect.Value.Elem({0xe4dec0?, 0xc000917c70?, 0x2?})
> 	/usr/local/go/src/reflect/value.go:1271 +0x45f
> github.com/apache/tinkerpop/gremlin-go/v3/driver/cucumber.getEdges(0xc000032053?)
> 	/go_app/driver/cucumber/cucumberWorld.go:180 +0xe99
> github.com/apache/tinkerpop/gremlin-go/v3/driver/cucumber.(*CucumberWorld).loadAllDataGraph(0xc0002bf4a0)
> 	/go_app/driver/cucumber/cucumberWorld.go:115 +0x3c9
> github.com/apache/tinkerpop/gremlin-go/v3/driver/cucumber.InitializeTestSuite.func1()
> 	/go_app/driver/cucumber/cucumberSteps_test.go:878 +0x45
> github.com/cucumber/godog.(*runner).concurrent(0xc000300a80, 0x1)
> 	/go/pkg/mod/github.com/cucumber/godog@v0.14.1/run.go:82 +0x602
> github.com/cucumber/godog.runWithOptions({_, _}, {0x0, 0x0, 0x0, {0x0, 0x0}, 0x0, {0x0, 0x0, ...}, ...}, ...)
> 	/go/pkg/mod/github.com/cucumber/godog@v0.14.1/run.go:284 +0x13ea
> github.com/cucumber/godog.TestSuite.Run({{0x0, 0x0}, 0xf40478, 0xf40460, 0xc0002ca0b0})
> 	/go/pkg/mod/github.com/cucumber/godog@v0.14.1/run.go:348 +0x2d4
> github.com/apache/tinkerpop/gremlin-go/v3/driver/cucumber.TestCucumberFeatures(0xc00015e9c0)
> 	/go_app/driver/cucumber/cucumberSteps_test.go:951 +0x347
> testing.tRunner(0xc00015e9c0, 0xf40300)
> 	/usr/local/go/src/testing/testing.go:1595 +0x262
> created by testing.(*T).Run in goroutine 1
> 	/usr/local/go/src/testing/testing.go:1648 +0x846
> ```
> 
> Also, you should add an entry to the CHANGELOG.asciidoc.
> 
> I'm not that familiar with Go, but I'll have another committer that is review this PR as well.

Thank you for your feedback and guidance.

I have addressed the issue and implemented the necessary fixes. Please feel free to run the Github Actions again to verify these changes.

Additionally, I have updated the CHANGELOG.asciidoc file with the relevant details pertaining to the changes made in this PR.

I appreciate your time and effort in reviewing this PR. I look forward to any further suggestions or feedback from the other committer who will review this.

Best Regards.

vkagamlyk commented:
VOTE+1

xiazcy commented:
Thank you for the improvement and added test cases, LGTM. VOTE +1

kenhuuu commented:
VOTE +1.

Please follow the instructions in the PR template next time for the CHANGELOG entry. I'll commit a follow-up change to make a minor adjustment so that this PR can be merged now. Thanks again for contributing.
